File: solutions/cpp/combinations.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <random>
#include <string>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::random_device;
using std::stoi;
using std::uniform_int_distribution;
using std::vector;

void DirectedCombinations(int, int,  int, vector<int>*, vector<vector<int>>*);

// @include
vector<vector<int>> Combinations(int n, int k) {
  vector<vector<int>> result;
  vector<int> partial_combination;
  DirectedCombinations(n, k, 1, &partial_combination, &result);
  return result;
}

void DirectedCombinations(int n, int k, int offset,
                          vector<int>* partial_combination,
                          vector<vector<int>>* result) {
  if (partial_combination->size() == k) {
    result->emplace_back(*partial_combination);
    return;
  }

  // Generate remaining combinations over {offset, ..., n - 1} of size
  // kNumRemaining.
  const int kNumRemaining = k - partial_combination->size();
  for (int i = offset; i <= n && kNumRemaining <= n - i - 1;
       ++i) {
    partial_combination->emplace_back(i);
    DirectedCombinations(n, k, i + 1, partial_combination, result);
    partial_combination->pop_back();
  }
}
// @exclude

int main(int argc, char** argv) {
  default_random_engine gen((random_device())());
  int n, k;
  if (argc == 3) {
    n = stoi(argv[1]), k = stoi(argv[2]);
  } else {
    uniform_int_distribution<int> n_dis(1, 10);
    n = n_dis(gen);
    uniform_int_distribution<int> k_dis(0, n);
    k = k_dis(gen);
  }
  auto result = Combinations(n, k);
  cout << "n = " << n << ", k = " << k << endl;
  for (const auto& vec : result) {
    for (int a : vec) {
      cout << a << " ";
    }
    cout << endl;
  }
  return 0;
}

File: solutions/python/combinations.py
import sys
import random


# @include
def combinations(n, k):
    def directed_combinations(offset, partial_combination):
        if len(partial_combination) == k:
            result.append(list(partial_combination))
            return

        # Generate remaining combinations over {offset, ..., n - 1} of size
        # num_remaining.
        num_remaining = k - len(partial_combination)
        i = offset
        while i <= n and num_remaining <= n - i + 1:
            directed_combinations(i + 1, partial_combination + [i])
            i += 1

    result = []
    directed_combinations(1, [])
    return result
# @exclude


def combinations_pythonic(n, k):
    res = [[]]
    for _ in range(k):
        res = [[i] + c for c in res for i in range(1, c[0] if c else n + 1)]
    return res


def small_test():
    assert combinations(4, 2) == [[1, 2], [1, 3], [1, 4], [2, 3], [2, 4],
                                  [3, 4]]
    assert combinations_pythonic(4, 2) == [[1, 2], [1, 3], [2, 3], [1, 4],
                                           [2, 4], [3, 4]]


def main():
    small_test()
    if len(sys.argv) == 3:
        n = int(sys.argv[1])
        k = int(sys.argv[2])
    else:
        n = random.randint(1, 10)
        k = random.randint(0, n)
    result = combinations(n, k)
    print('n = ', n, ', k = ', k, sep='')
    for vec in result:
        print(*vec)


if __name__ == '__main__':
    main()
